chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/state/JoinedRoomsStateDiffer.kt

```diff
@@ -0,0 +1,86 @@
+package com.pusher.chatkit.rooms.state
+
+import com.pusher.chatkit.state.Action
+import com.pusher.chatkit.state.JoinedRoom
+import com.pusher.chatkit.state.LeftRoom
+import com.pusher.chatkit.state.RoomUpdated
+import com.pusher.chatkit.state.State
+import org.reduxkotlin.GetState
+
+internal class JoinedRoomsStateDiffer(val stateGetter: GetState<State>) {
+
+    internal fun stateExists() = stateGetter().joinedRoomsState != null
+
+    internal fun toActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> {
+        return joinedRoomActions(newRooms, newUnreadCounts) +
+            roomUpdatedActions(newRooms) +
+            leftRoomActions(newRooms)
+    }
+
+    private val currentRooms get() = stateGetter().joinedRoomsState!!.rooms
+
+    private fun joinedRoomActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> {
+        val actions = mutableListOf<Action>()
+
+        val addedRoomKeys =
+            newRooms.map { it.id }.toSet() - currentRooms.keys.toSet()
+
+        for (addedKey in addedRoomKeys) {
+            val room = newRooms.find { it.id == addedKey }
+            val unreadCount = newUnreadCounts[addedKey]
+
+            if (room != null &&
+                unreadCount != null
```

Comment:
That's not correct to drop rooms missing unread count. We should have a separate event `ReconnectJoinedRoom` which has `unreadCount` as optional.

Reply:
hmm, i thought this was addressed [here](https://github.com/pusher/chatkit-android/pull/193#issuecomment-593962906) as part of my initial questions. Have updated as per your comments below too.